src/java/org/apache/cassandra/service/ConsensusTableMigrationState.java

```diff
@@ -0,0 +1,697 @@
+                Set<Range<Token>> migratedRanges = deserializeSet(in, partitioner, version, rangeSerializer);
+                Map<Epoch, List<Range<Token>>> migratingRangesByEpoch = deserializeMap(in, partitioner, version, Epoch.messageSerializer, rangesSerializer, newHashMap());
+                return new TableMigrationState(keyspaceName, tableName, tableId, targetProtocol, migratedRanges, migratingRangesByEpoch);
+            }
+
+            @Override
+            public long serializedSize(TableMigrationState t, int version)
+            {
+                return sizeof(t.targetProtocol.value)
+                        + sizeof(t.keyspaceName)
+                        + sizeof(t.tableName)
+                        + t.tableId.serializedSize()
+                        + serializedCollectionSize(t.migratedRanges, version, rangeSerializer)
+                        + serializedMapSize(t.migratingRangesByEpoch, version, Epoch.messageSerializer, rangesSerializer);
+            }
+        };
+
+        @Nonnull
+        public final List<Range<Token>> migratingRanges;
+
+        @Nonnull
+        public final List<Range<Token>> migratingAndMigratedRanges;
+
+        public TableMigrationState(@Nonnull String keyspaceName,
+                                   @Nonnull String tableName,
+                                   @Nonnull TableId tableId,
+                                   @Nonnull ConsensusMigrationTarget targetProtocol,
+                                   @Nonnull Collection<Range<Token>> migratedRanges,
+                                   @Nonnull Map<Epoch, List<Range<Token>>> migratingRangesByEpoch)
+        {
+            this.keyspaceName = keyspaceName;
+            this.tableName = tableName;
+            this.tableId = tableId;
+            this.targetProtocol = targetProtocol;
+            this.migratedRanges = ImmutableList.copyOf(normalize(migratedRanges));
+            this.migratingRangesByEpoch = ImmutableSortedMap.copyOf(
+            migratingRangesByEpoch.entrySet()
+                                  .stream()
+                                  .map( entry -> new SimpleEntry<>(entry.getKey(), ImmutableList.copyOf(normalize(entry.getValue()))))
+                                  .collect(Collectors.toList()));
+            this.migratingRanges = ImmutableList.copyOf(normalize(migratingRangesByEpoch.values().stream().flatMap(Collection::stream).collect(Collectors.toList())));
+            this.migratingAndMigratedRanges = ImmutableList.copyOf(normalize(ImmutableList.<Range<Token>>builder().addAll(migratedRanges).addAll(migratingRanges).build()));
+        }
+
+        public TableMigrationState withRangesMigrating(@Nonnull Collection<Range<Token>> ranges,
+                                                       @Nonnull ConsensusMigrationTarget target,
+                                                       @Nonnull Epoch epoch)
+        {
+            checkArgument(epoch.isAfter(Epoch.EMPTY), "Epoch shouldn't be empty");
```

Comment:
should validation be more tolerant of overlapping/duplicate begin migration commands? I don't think there's any real downside to logging that the range has already been migrated or is migrating so the command will be either be a noop or will affect a subset of the specified ranges. I'd imagine this will make life easier for operators at least.

Reply:
I agree. I'll convert the errors to logging.